stream_alert/rule_processor/rules_engine.py

```diff
@@ -403,3 +419,36 @@ def process(cls, input_payload):
                     alerts.append(alert)
 
         return alerts
+
+    @classmethod
+    def get_intelligence(cls, intel_dir='threat_intel'):
+        """Load intelligence from csv.gz files into a dictionary
+
+        Args:
+            intel_dir (str): Location where stores compressed intelligence
+        """
+        threat_intel = ThreatIntel(intel_dir)
+        if cls.__intelligence:
+            return
+        cls.__intelligence = threat_intel.read_compressed_files()
+
+    @classmethod
+    def is_ioc(cls, rec):
+        """Detect if a record contains any data matching to IOC (Indicate of
+        compromise) which store in class variable intelligence dictionary.
+
+        Args:
+            record (dict): Payload record to process
+
+        Return:
+            (tuple): (True or False, IOC type, IOC value)
+        """
+        _, datatypes_ioc_mapping = load_threat_intel_conf()
+        for datatype in rec['normalized_types']:
```

Comment:
* `is_ioc` function is only called when there is `datatypes` defined in a rule.
* `rec` argument is a parsed and "half-processed" record, there is `normalized_types` key but its value could be empty when this record is passed to `is_ioc` function.

Reply:
I'm pretty confused. Is the idea that

1) `is_ioc()` will be called from rules themselves, or

2) `is_ioc()` is handled internally within the processor, and it's just the responsibility of the user to define types and their mappings in JSON files and it just magically works

3) both (1) and (2)?

From your implementation, it seems like you went the route of `(2)`, since the only `is_ioc()` call can be found in `rules_engine.py`, and, it doesn't accept arguments that would support a use-case like `is_ioc(record['foobar'], IOC_TYPE::DOMAIN)` from within a rule.

If this is the case, your first bullet in your response to Jack is confusing. Why would `datatypes` need to be set in a rule in order for this to work?

Please clarify

Reply:
@mime-frame route of (2) is the one I implemented. `is_ioc()` is handled internally, and it will not called from rules themselves. Users need to define types and their mappings in the JSON files as you mentioned in route (2). We can improve the user experience in near future. 

Currently, it doesn't support user-case like `is_ioc(record['foobar'], IOC_TYPE::DOMAIN)` within a rule.

Reply:
Which means this comment:

> is_ioc function is only called when there is datatypes defined in a rule.

Is factually incorrect

Reply:
Just wanted to ensure we're all on the same page.

Reply:
> is_ioc function is only called when there is datatypes defined in a rule.

I probably should reword this comment to `is_ioc function is working with Data Normalization currently.`
In current implementation stage, Threat Intel is strongly tight with Data Normalization. Without Data Normalization we have to define a bigger map between original keys in each schema and IOC types. 

Does it make sense to you?